pkg/splunk/enterprise/workerpool.go

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2018-2021 Splunk Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package enterprise
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	enterpriseApi "github.com/splunk/splunk-operator/pkg/apis/enterprise/v2"
+	splclient "github.com/splunk/splunk-operator/pkg/splunk/client"
+)
+
+// AppInstallWorkerPool is a pool of workers to install apps in parallel
+type AppInstallWorkerPool struct {
+	NumOfWorkers       uint64
+	JobList            []*enterpriseApi.AppDeploymentInfo
+	JobsChan           chan *enterpriseApi.AppDeploymentInfo
+	DownloadPath       string
+	Scope              string
+	S3ClientMgr        S3ClientManager
+	S3Reponse          splclient.S3Response
+	wg                 *sync.WaitGroup
+	availableDiskSpace uint64
+	mu                 sync.Mutex
+}
+
+// NewAppInstallWorkerPool returns a new worker pool
+func NewAppInstallWorkerPool(numOfWorkers uint64, jobs []enterpriseApi.AppDeploymentInfo,
+	s3ClientMgr S3ClientManager, s3Response splclient.S3Response,
+	availableDiskSpace uint64, downloadPath, scope string) *AppInstallWorkerPool {
+	w := &AppInstallWorkerPool{
+		NumOfWorkers:       numOfWorkers,
+		JobList:            getObjectsAsPointers(jobs).([]*enterpriseApi.AppDeploymentInfo),
+		JobsChan:           make(chan *enterpriseApi.AppDeploymentInfo),
+		DownloadPath:       downloadPath,
+		Scope:              scope,
+		S3ClientMgr:        s3ClientMgr,
+		S3Reponse:          s3Response,
+		availableDiskSpace: availableDiskSpace,
+		wg:                 new(sync.WaitGroup),
+	}
+	return w
+}
+
+// Run is the main work that workers do to install the apps.
```

Comment:
Am I right in saying the main work is to download the apps and not install?

Reply:
As of now, it's download, but we plan to extend it to copy and other stuff too. If we plan to create separate threads for other tasks, then I can rename it to suggest it is for Download only.